### PR TITLE
Low power management module for PIC32

### PR DESCRIPTION
**Pull request by ghost**
This connects to #29.

Implement power management for I2C, SPI and UART.

### CODE REVIEW

File: platform/mikro-e/contiki-mikro-e-main.c [hunk 1]
```diff
@@ -62,7 +65,36 @@ SENSORS(&button_sensor, &button_sensor2, &proximity_sensor);
 #else
 SENSORS(&button_sensor, &button_sensor2);
 #endif
+/*---------------------------------------------------------------------------*/
+static void power_down_peripherals(void)
+{
+  #ifdef __USE_SPI_PORT1__
+  void pic32_spi1_power_down();
```
Comment (Avinash95) [inline platform/mikro-e/contiki-mikro-e-main.c]: void ?

File: cpu/pic32/lib/pic32_i2c.h [hunk 2]
```diff
@@ -42,6 +42,8 @@
 #ifdef __USE_I2C__
 /*---------------------------------------------------------------------------*/
 #define I2C_DEF(XX)                                          \
+  void pic32_i2c##XX##_power_down(void);                           \
```
Comment (mtusnio) [inline cpu/pic32/lib/pic32_i2c.h]: By default all driver functions always return something, even if it's always 0.

File: cpu/pic32/lib/pic32_uart.h [hunk 3]
```diff
@@ -76,6 +76,8 @@
 #define UART_ERR_OVERFLOW          2
 
 #define UART_PORT_DEF(XX)                                                \
+  void pic32_uart##XX##_power_down(void);                                \
```
Comment (mtusnio) [inline cpu/pic32/lib/pic32_uart.h]: Add a return type

**Review by mtusnio — CHANGES_REQUESTED** [review]

File: cpu/pic32/lib/pic32_i2c.c [hunk 4]
```diff
@@ -53,8 +53,32 @@
...
+  {                                                                      \
+    /* Save copy of register I2C##XX##CON to restore its state           \
+     * while powering up the peripheral.                                 \
+     */                                                                  \
+    pic32_i2c##XX##_state = I2C##XX##CON;                                \
+    I2C##XX##CONCLR = _I2C##XX##CON_ON_MASK;                             \
+    PMD5SET = _PMD5_I2C##XX##MD_MASK;                                    \
+    return 0;                                                            \
+  }                                                                      \
+  int8_t                                                                 \
```
Comment (Avinash95) [inline cpu/pic32/lib/pic32_i2c.c:78]: there is no error handling in the function, should return type be void ?

File: cpu/pic32/lpm.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+}
+/*---------------------------------------------------------------------------*/
+void
+lpm_enter(void)
+{
+  power_down_peripherals();
+
+  asm volatile("wait");
```
Comment (Avinash95) [inline cpu/pic32/lpm.c:83]: is this instruction aimed at to put the CPU in the idle mode ?

**Review by Avinash95 — DISMISSED** [review]
we are using timer1 and timer23, should we disable them as well ?

File: cpu/pic32/lpm.c (hunk 5 above)
Comment (ghost) [reply cpu/pic32/lpm.c:83]: Yes, idle or sleep mode, it depends on the value of SLPEN bit.

**Comment by francois-berder-imgtec** [comment]
@Avinash95 timer2/3 are used for rtimer, and we need it to be able to wake-up the cpu in order to execute a task at a regular interval. Same for timer1 controlling the etimer.

File: cpu/pic32/lib/pic32_i2c.c (hunk 4 above)
Comment (ghost) [reply cpu/pic32/lib/pic32_i2c.c:78]: No, I think it should still return ```int8_t``` to avoid changing the API in case we need to handle an error.

**Review by mtusnio — CHANGES_REQUESTED** [review]
SPI doesn't work for LED Matrix

**Comment by ghost** [comment]
I added some code to make LPM optional.

**Comment by ghost** [comment]
This pr is blocked on #69 .

**Comment by ghost** [comment]
@mtusnio It appears that disabling spi module used by cascoda and uart module used for debug output make the clicker impossible to wake-up. I think the clicker is probably stuck in an infinite loop.
As a consequence, I changed my code to enable LPM per module.

**Review by mtusnio — CHANGES_REQUESTED** [review]
SPI mode changes between power down/power up, UART does not work.

**Review by mtusnio — APPROVED** [review]

**Comment by mtusnio** [comment]
@Avinash95 can you review and approve if you are happy with the PR?

**Pull request merged**